src/cluster_import.c

```diff
@@ -0,0 +1,1927 @@
+        addReplyErrorFormat(c, "Link name not found");
+        return;
+    }
+    if (link->one_shot) {
+        addReplyErrorFormat(c, "Link is executing in one-shot mode");
+        return;
+    }
+    if (link->state != SLOT_IMPORT_RECEIVE_STREAM) {
+        addReplyErrorFormat(c, "Link is not ready to be committed");
+        return;
+    }
+
+    /* Start the failover procedure */
+    sendSyncSlotsMessageOnLink(link, "PAUSE");
+    updateSlotMigrationLinkState(link, SLOT_IMPORT_FAILOVER_WAITING_FOR_PAUSED);
+
+    /* If there was a status from a previous attempt, clear it as this is a fresh attempt. */
+    if (link->status_msg) {
+        sdsfree(link->status_msg);
+        link->status_msg = NULL;
+    }
+    addReply(c, shared.ok);
+}
+
+/* Cancels one or all ongoing imports. */
+void clusterCommandImportCancel(client *c) {
+    listNode *ln;
+    listIter li;
+
+    if (!strcasecmp(c->argv[2]->ptr, "all")) {
+        if (!clusterIsAnySlotImporting()) {
+            addReplyError(c, "No imports ongoing");
+            return;
+        }
+
+        listRewind(server.cluster->slot_migration_links, &li);
+        while ((ln = listNext(&li)) != NULL) {
+            slotMigrationLink *link = ln->value;
+            if (!isSlotMigrationLinkInProgress(link)) {
+                continue;
+            }
+            if (link->client) {
+                sendSyncSlotsMessageOnLink(link, "CANCEL");
+                link->client->flag.close_after_reply = 1;
+            }
+            finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_CANCELLED, NULL);
+        }
+        serverLog(LL_NOTICE, "Canceled all in progress slot imports due to operator's request.");
+        addReply(c, shared.ok);
+        return;
+    } else if (!strcasecmp(c->argv[2]->ptr, "link") && c->argc > 3) {
+        slotMigrationLink *link = clusterLookupImportLink(c->argv[3]->ptr);
+        if (!link || !isSlotMigrationLinkInProgress(link)) {
+            addReplyErrorFormat(c, "Link name not found.");
+            return;
+        }
+        if (link->client) {
+            sendSyncSlotsMessageOnLink(link, "CANCEL");
+            link->client->flag.close_after_reply = 1;
+        }
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_CANCELLED, NULL);
+        serverLog(LL_NOTICE,
+                  "Slot import link %.40s to node %.40s (owner of slots [%s]) "
+                  "cancelled due to operator's request.",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        addReply(c, shared.ok);
+        return;
+    } else {
+        addReplyErrorObject(c, shared.syntaxerr);
+        return;
+    }
+}
+
+/* Sent by the source to the target after dumping the snapshot in
+ * AOF format. */
+void clusterCommandSyncSlotsSnapshotEof(client *c) {
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    if (!link || link->type != SLOT_MIGRATION_IMPORT) {
+        serverLog(LL_WARNING,
```

Comment:
I don't think we really need to log this, we should probably return an error to the client and not just kill it as well. If there is some issue in the state machine, we'll just get a disconnect and not a useful error.

Reply:
> I guess we could also reply with an error, which will produce a log due to the networking.c code that checks that responses aren't sent to the slot migration source.

This is what I was expecting and makes it a bit easier to debug since you get the error. I suppose I don't feel super strongly one way or the other. It also makes it easier for any type of interoperability system to get a coherent error back if they are trying to interact with it. We (AWS) have one like that for replication but not for slot migration, but in principal a system could try to emulate either side of the slot migration.

Reply:
I've now added this - so we will send:

CLUSTER SYNCSLOTS ESTABLISH ...

And we will enter a state to read out the response before we promote the connection to a client.

It does add some complexity (any time you need to do async IO it takes some code and debugging) but the end result does feel cleaner (and now I can get rid of the additional SYNCSLOTS FAIL command)